independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/SyntheticCreationalContextImpl.java

```diff
@@ -0,0 +1,102 @@
+package io.quarkus.arc.impl;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Objects;
+
+import jakarta.enterprise.context.spi.CreationalContext;
+import jakarta.enterprise.inject.Default;
+import jakarta.enterprise.util.TypeLiteral;
+
+import io.quarkus.arc.BeanCreator.SyntheticCreationalContext;
+
+public final class SyntheticCreationalContextImpl<T> implements SyntheticCreationalContext<T> {
+
+    private final CreationalContext<T> creationalContext;
+    private final Map<TypeAndQualifiers, Object> injectedReferences;
+    private final Map<String, Object> params;
+
+    public SyntheticCreationalContextImpl(CreationalContext<T> creationalContext, Map<String, Object> params,
+            Map<TypeAndQualifiers, Object> injectedReferences) {
+        this.creationalContext = Objects.requireNonNull(creationalContext);
+        this.params = Objects.requireNonNull(params);
+        this.injectedReferences = Objects.requireNonNull(injectedReferences);
+    }
+
+    @Override
+    public void push(T incompleteInstance) {
+        creationalContext.push(incompleteInstance);
+    }
+
+    @Override
+    public void release() {
+        creationalContext.release();
+    }
+
+    @Override
+    public Map<String, Object> getParams() {
+        return params;
+    }
+
+    @Override
+    public <R> R getInjectedReference(Class<R> requiredType, Annotation... qualifiers) {
+        return getReference(requiredType, qualifiers);
+    }
+
+    @Override
+    public <R> R getInjectedReference(TypeLiteral<R> requiredType, Annotation... qualifiers) {
+        return getReference(requiredType.getType(), qualifiers);
+    }
+
+    @SuppressWarnings("unchecked")
+    private <R> R getReference(Type requiredType, Annotation... qualifiers) {
+        if (qualifiers.length == 0) {
+            qualifiers = new Annotation[] { Default.Literal.INSTANCE };
+        }
+        R ref = (R) injectedReferences.get(new TypeAndQualifiers(requiredType, qualifiers));
+        if (ref == null) {
+            throw new IllegalArgumentException("A synthetic injection point was not declared for required type [" + requiredType
+                    + " and qualifiers: " + Arrays.toString(qualifiers));
+        }
+        return ref;
+    }
+
+    public static class TypeAndQualifiers {
```

Comment:
I'd make this a top-level class.

Reply:
:-1: this class belongs to the `SyntheticCreationalContextImpl`.

Reply:
I understand this is an implementation detail of `SyntheticCreationalContextImpl`, but these are all `public`, and we're bound to find a use for this outside of synthetic beans sooner or later :-) But your call.

Reply:
Then we'll extract the class when it's needed... ;-)